protocol/op-supervisor-integration.md

```diff
@@ -0,0 +1,309 @@
+    - If the entry is not yet known, return an `unknown` safety level.
+
+
+### `TryCrossUnsafe`
+
+```go
+type RPC interface {
+    GetCrossUnsafe(chainID types.ChainID, maxNumber uint64) (crossUnsafe eth.BlockID, err error)
+}
+```
+
+API implementation:
+1. Compare the `maxNumber` to the locally known `unsafe` tip of the event-DB:
+    - If newer:
+      1. Initiate fetching of new event-log data.
+      2. Return the latest known `cross-unsafe`.
+    - If equal or older:
+      1. Return the latest known `cross-unsafe` block before `maxNumber`.
+         This is the last block checkpoint `<= maxNumber`
+
+By clipping using `maxNumber` the op-node is able to verify the returned `cross-unsafe`, and not accept it if it's non-canonical, and optionally query for older cross-unsafe blocks to try to bump to.
+
+### `NextDeriveTask`, `OnDerived`, `CrossSafe`
+
+```go
+type RPC interface {
+	
+	// Returns where the next L2 block has to be derived from. 
+	NextDeriveTask(chainID types.ChainID) (parentL2 eth.BlockID, depositOnly bool, l1 eth.BlockID) error
+
+	// Signals what has been derived, considered to be local-safe in op-node.
+	OnDerived(chainID types.ChainID, derived eth.BlockID, parentL2 eth.BlockID, depositOnly bool, l1 eth.BlockID) error
+
+	// CrossSafe returns the cross-safe block, and where in L1 traversal it was recognized as cross-safe. 
+	CrossSafe(chainID types.ChainID, l1Number uint64) (crossSafe eth.BlockID, derivedFrom eth.BlockID, error)
+}
+```
+`NextDeriveTask`: Given the chainID, check if we have a local-safe block that builds on top of the latest cross-unsafe.
+- If we do not have such block, then create the task to derive one:
+  - `{parentL2: latestCrossSafe, depositOnly: false, l1: l1OfLatestCrossSafe}`
+- If we do, then determine if the local-safe block has been verified:
+  1. Check if all dependencies are:
+    - known
+    - cross-safe if previous height
+    - local-safe if same height
+    - have valid intra-block dependencies if same height
+  2. If unknowns, then return error to defer the task till later.
+  3. If invalid dependencies, then create a task to replace the block:
+    - `{parentL2: latestCrossSafe, depositOnly: true, l1: l1OfLatestCrossSafe}`
+
+`OnDerived`: the op-node signals the result of a previous `NextDeriveTask`:
+- If `parentL2`, `depositOnly`, `l1` match the current task:
+    1. Accept it, and merge it into the `super-safe-db` head entry.
+    2. The head-entry should be flushed to the DB, and tasks should be updated, once every L2 chain has been seen.
+       - TODO: multiple L2 blocks from the same L1 block. We need some way to report exhaustion of a L1 block.
+       - TODO: there may not be any L2 blocks derived from the L1 block. Can repeat parentL2 as "derived" in that case maybe?
+- If it does not match the task, then heck if it matches a past entry.
+  - If it does, silently accept the result. There may be multiple op-nodes running the same tasks.
```

Comment:
ooo, multiple op-nodes driving this makes things very very interesting.   We might be getting `OnDerived` or `GetCrossUnsafe` calls from a different op-node to the one that we're trying to pull logs from when updating our unsafe head.  Having a single shared op-supervisor is a pretty huge win if we can do it, but lots of complexity in it when its controlling derivation tasks.

Reply:
- If the supervisor task is ahead of an op-node, then the op-node can ignore it.
- If the supervisor task is behind of an op-node, then the op-node should reset back (can be a separate different derivation-pipeline process) to help determine the L1->L2 mapping.
- If the supervisor task is equal to what the op-node does, then the op-node can process it, and publish back to the supervisor. In the meantime the supervisor may have gotten behind (reorg) / ahead (through other op-node), but should be able to check consistency with where it is.

When multiple op-nodes of the same L2 are helping the supervisor, then at random the supervisor will get ahead through the fastest op-node, and ignore the duplicate work of others (if performed at all).